File: synthlog/clause.py
from __future__ import print_function

from problog import get_evaluatable
from problog.extern import problog_export, problog_export_nondet, problog_export_raw

from problog.logic import Clause, Constant, Term
from copy import deepcopy

existing_terms = set()


@problog_export_raw("+term")
def store_existing_term(term, *args, **kwargs):
    global existing_terms
    existing_terms.add(term)
    return []


@problog_export_raw("+term")
def is_not_existing_term(term, *args, **kwargs):
    global existing_terms
    res = []
    if term not in existing_terms:
        res.append((term,))
    return res


@problog_export_raw("+term")
def standard_term(term, *args, **kwargs):
    if term.functor in ("\\+", "not"):
        return []
    return [(term,)]


@problog_export_nondet("+term", "-term", "-term", "-term")
def unify_clause(clause):
    if isinstance(clause, Clause):
        head = deepcopy(clause.head)
        proba = head.probability
        if proba is None:
            proba = 1
        head.probability = None
        if not isinstance(proba, Constant):
            proba = Constant(proba)
        return [(head, Constant(proba), clause.body)]
    return []


@problog_export("+term", "+term")
def apply_clause(scope, clause, database=None, *args, **kwargs):
    if isinstance(clause, Clause):
        database.add_clause(
            Clause(
                Term(
                    "':'",
                    scope,
                    clause.head.with_probability(None),
                    p=clause.head.probability,
                ),
                Term("evaluate", scope, clause.body),
            )
        )
    return ()

File: synthlog/clause.pl
:- use_module('clause.py').

% Clause parsing
:- Scope:Z, apply_clause(Scope,Z).
%P::Scope:X :- contains_clauses(Scope), Scope:Z, Z \== X, unify_clause(Z,X,P,Y), evaluate(Scope,Y).

contains_clauses(dummyscopeble).

% Evaluation rules
evaluate(Scope,Y) :- Scope:Y.
evaluate(Scope,var(Y)) :- Scope:Y.
%% \+ Y
evaluate(Scope, \+ Y) :- \+ evaluate(Scope, Y). % (evaluate(Scope,Y), store_existing_term(Y)); is_not_existing_term(Y).
evaluate(Scope, not(Y)) :- evaluate(Scope, \+ Y).
%% X < Y
evaluate(Scope,(X < Y)) :- ground(X), ground(Y), X < Y.
%% X <= Y
evaluate(Scope,(X =< Y)) :- ground(X), ground(Y), X =< Y.
%% X > Y
evaluate(Scope,(X > Y)) :- ground(X), ground(Y), X > Y.
%% X >= Y
evaluate(Scope,(X >= Y)) :- ground(X), ground(Y), X >= Y.
%% X == Y
evaluate(Scope, (X == Y)) :- ground(X), X == Y.
evaluate(Scope, (X == Y)) :- ground(Y), X == Y.
%% X \== Y
evaluate(Scope, (X \== Y)) :- ground(X), ground(Y), X \== Y.
%% X is Y
evaluate(Scope,(X is Y)) :- ground(Y), X is Y.
%% X, Y
evaluate(Scope,(X,Y)) :- evaluate(Scope,X), evaluate(Scope,Y).
evaluate(Scope,(X,Y)) :- evaluate(Scope,Y), evaluate(Scope,X).
%% X; Y
evaluate(Scope,(X;Y)) :- evaluate(Scope,X); evaluate(Scope,Y).
evaluate(Scope,(X;Y)) :- evaluate(Scope,Y); evaluate(Scope,X).